### PR TITLE
`azurerm_private_endpoint` - expose `private_dns_zone_group`, `private_dns_zone_configs`, and `custom_dns_configs`

### LINKED ISSUE #6571: Support for list of private_ip_address attributes on azurerm_private_endpoint resource
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

The private endpoint for Azure Container Registry exposes two IP addresses, both of which are required for using ACR.  The `azurerm_private_endpoint` should expose both, but right now, it only exposes the one, via `private_ip_address`.

### New or Affected Resource(s)

* azurerm_private_endpoint

### Potential Terraform Configuration

```hcl
resource "azurerm_resource_group" "example" {
  name     = "example-resources"
  location = "West Europe"
}

resource "azurerm_virtual_network" "example" {
  name                = "example-network"
  address_space       = ["10.0.0.0/16"]
  location            = azurerm_resource_group.example.location
  resource_group_name = azurerm_resource_group.example.name
}

resource "azurerm_subnet" "endpoint" {
  name                 = "endpoint"
  resource_group_name  = azurerm_resource_group.example.name
  virtual_network_name = azurerm_virtual_network.example.name
  address_prefix       = "10.0.2.0/24"

  enforce_private_link_endpoint_network_policies = true
}

resource "azurerm_container_registry" "example" {
  name                     = "example-containerregistry"
  resource_group_name      = azurerm_resource_group.example.name
  location                 = azurerm_resource_group.example.location
  sku                      = "Premium"
  admin_enabled            = false
}

resource "azurerm_private_endpoint" "example" {
  name                = "example-endpoint"
  location            = azurerm_resource_group.example.location
  resource_group_name = azurerm_resource_group.example.name
  subnet_id           = azurerm_subnet.endpoint.id

  private_service_connection {
    name                           = "example-privateserviceconnection"
    private_connection_resource_id = azurerm_container_registry.example.id
    subresource_names              = ["registry"]
    is_manual_connection           = false
  }
}

output acr_data_private_ip {
  value = azurerm_private_endpoint.example.private_ip_address[0]
}

output acr_registry_private_ip {
  value = azurerm_private_endpoint.example.private_ip_address[1]
}
```

### References

* https://github.com/terraform-providers/terraform-provider-azurerm/issues/6146#issuecomment-601899311
* https://docs.microsoft.com/en-us/azure/container-registry/container-registry-private-link#get-private-ip-addresses

### PR DESCRIPTION
**Pull request by WodansSon**
(fixes #6571)

### CODE REVIEW

**Comment by WodansSon** [comment]
image

File: azurerm/internal/services/network/private_endpoint_resource.go [hunk 1]
```diff
@@ -148,30 +251,107 @@ func resourceArmPrivateEndpointCreateUpdate(d *schema.ResourceData, meta interfa
...
-		return fmt.Errorf("Error retrieving Private Endpoint %q (Resource Group %q): %+v", name, resourceGroup, err)
+		return fmt.Errorf("retrieving Private Endpoint %q (Resource Group %q): %+v", name, resourceGroup, err)
 	}
 	if resp.ID == nil || *resp.ID == "" {
 		return fmt.Errorf("API returns a nil/empty id on Private Endpoint %q (Resource Group %q): %+v", name, resourceGroup, err)
 	}
 	d.SetId(*resp.ID)
 
+	// now create the dns zone group
+	// first I have to see if the dns zone group exists, if it does I need to delete it an re-create it because you can only have one per private endpoint
```
Comment (tombuildsstuff) [inline azurerm/internal/services/network/private_endpoint_resource.go:273]: what happens to traffic when this is deleted, is this dropped? if so - should we be locating the existing one and updating instead?

Comment (WodansSon) [reply azurerm/internal/services/network/private_endpoint_resource.go:272]: I only delete if the name has changed, else I will need to force new the attribute which would delete the entire private endpoint.

File: azurerm/internal/services/network/parse/private_endpoint.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+	id, err := azure.ParseAzureResourceID(input)
+	if err != nil {
+		return NameResourceGroup{}, fmt.Errorf("unable to parse Private DNS Zone Group ID %q: %+v", input, err)
+	}
+
+	privateDnsZoneGroup := NameResourceGroup{
+		Name:          id.Path["privateDnsZoneGroups"],
+		ResourceGroup: id.ResourceGroup,
+		ID:            input,
+	}
```
[the remark this answers is not in the document]
Comment (WodansSon) [reply azurerm/internal/services/network/parse/private_endpoint.go:30]: Fixed.

File: azurerm/internal/services/network/parse/private_endpoint.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+			ID:            v,
+		}
+
+		results = append(results, privateDnsZone)
+	}
+
+	return results, nil
+}
+
+func PrivateEndpointResourceID(input string) (NameResourceGroup, error) {
```
[the remark this answers is not in the document]
Comment (WodansSon) [reply azurerm/internal/services/network/parse/private_endpoint.go:58]: Fixed.

File: azurerm/internal/services/network/parse/private_endpoint.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+		id, err := azure.ParseAzureResourceID(v)
+		if err != nil {
+			return nil, fmt.Errorf("unable to parse Private DNS Zone ID %q: %+v", input, err)
+		}
+
+		privateDnsZone := NameResourceGroup{
+			Name:          id.Path["privateDnsZones"],
+			ResourceGroup: id.ResourceGroup,
+			ID:            v,
+		}
```
[the remark this answers is not in the document]
Comment (WodansSon) [reply azurerm/internal/services/network/parse/private_endpoint.go:50]: Fixed.

File: azurerm/internal/services/network/parse/private_endpoint.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+	for _, item := range input {
+		v := item.(string)
+
+		id, err := azure.ParseAzureResourceID(v)
+		if err != nil {
+			return nil, fmt.Errorf("unable to parse Private DNS Zone ID %q: %+v", input, err)
+		}
+
+		privateDnsZone := NameResourceGroup{
+			Name:          id.Path["privateDnsZones"],
```
[the remark this answers is not in the document]
Comment (WodansSon) [reply azurerm/internal/services/network/parse/private_endpoint.go:47]: Fixed.

File: azurerm/internal/services/network/private_endpoint_resource.go [hunk 6]
```diff
@@ -54,6 +55,33 @@ func resourceArmPrivateEndpoint() *schema.Resource {
...
+							Type:         schema.TypeString,
+							Required:     true,
+							ValidateFunc: ValidatePrivateLinkName,
+						},
+						"private_dns_zone_ids": {
+							Type:     schema.TypeList,
+							Required: true,
+							Elem: &schema.Schema{
+								Type:         schema.TypeString,
+								ValidateFunc: azure.ValidateResourceID,
```
[the remark this answers is not in the document]
Comment (WodansSon) [reply azurerm/internal/services/network/private_endpoint_resource.go:78]: Fixed.

**Comment by WodansSon** [comment]
=== RUN   TestAccAzureRMPrivateEndpoint_basic
=== PAUSE TestAccAzureRMPrivateEndpoint_basic
=== RUN   TestAccAzureRMPrivateEndpoint_updateTag
=== PAUSE TestAccAzureRMPrivateEndpoint_updateTag
=== RUN   TestAccAzureRMPrivateEndpoint_requestMessage
=== PAUSE TestAccAzureRMPrivateEndpoint_requestMessage
=== RUN   TestAccAzureRMPrivateEndpoint_privateDnsZoneGroup
=== PAUSE TestAccAzureRMPrivateEndpoint_privateDnsZoneGroup
=== RUN   TestAccAzureRMPrivateEndpoint_privateDnsZoneRename
=== PAUSE TestAccAzureRMPrivateEndpoint_privateDnsZoneRename
=== RUN   TestAccAzureRMPrivateEndpoint_privateDnsZoneUpdate
=== PAUSE TestAccAzureRMPrivateEndpoint_privateDnsZoneUpdate
=== RUN   TestAccAzureRMPrivateEndpoint_privateDnsZoneRemove
=== PAUSE TestAccAzureRMPrivateEndpoint_privateDnsZoneRemove
=== CONT  TestAccAzureRMPrivateEndpoint_basic
=== CONT  TestAccAzureRMPrivateEndpoint_privateDnsZoneRename
=== CONT  TestAccAzureRMPrivateEndpoint_privateDnsZoneRemove
=== CONT  TestAccAzureRMPrivateEndpoint_privateDnsZoneUpdate
=== CONT  TestAccAzureRMPrivateEndpoint_requestMessage
=== CONT  TestAccAzureRMPrivateEndpoint_privateDnsZoneGroup
=== CONT  TestAccAzureRMPrivateEndpoint_updateTag
--- PASS: TestAccAzureRMPrivateEndpoint_basic (179.87s)
--- PASS: TestAccAzureRMPrivateEndpoint_updateTag (237.52s)
--- PASS: TestAccAzureRMPrivateEndpoint_requestMessage (255.87s)
--- PASS: TestAccAzureRMPrivateEndpoint_privateDnsZoneGroup (313.20s)
--- PASS: TestAccAzureRMPrivateEndpoint_privateDnsZoneRename (362.99s)
--- PASS: TestAccAzureRMPrivateEndpoint_privateDnsZoneRemove (390.92s)
--- PASS: TestAccAzureRMPrivateEndpoint_privateDnsZoneUpdate (485.42s)
PASS
ok      github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/network/tests       485.510s

File: azurerm/internal/services/network/private_endpoint_resource.go [hunk 7]
```diff
@@ -286,6 +542,120 @@ func expandArmPrivateLinkEndpointServiceConnection(input []interface{}, parseMan
 	return &results
 }
 
+func flattenArmPrivateDnsZoneGroup(input network.PrivateDNSZoneGroup) []interface{} {
+	output := make([]interface{}, 0)
+
+	result := make(map[string]interface{})
+
+	result["id"] = *input.ID
+	result["name"] = *input.Name
```
Comment (katbyte) [inline azurerm/internal/services/network/private_endpoint_resource.go:551]: We should nil check these?

**Comment by WodansSon** [comment]
image

**Pull request merged**